### ISSUE TITLE
ROMIO w/ GPFS deadlocks

### ISSUE THREAD

**Issue by dbroemmel** [pmodels/mpich#3736]
When the GPFS filesystem is enabled, ROMIO deadlocks in a collective read operation. I presume it is not limited to read operations.

As far as I can tell it happens when the collective operation tries to use aggregators (maybe it also requires deferred open to be enabled). Then ranks behave differently depending on whether they are an aggregator or not, ending up with different calls to collectives inside `ad_gpfs_rdcoll.c`, e.g. in different branches of the `if` condition at https://github.com/pmodels/mpich/blob/2135f96a0580148563f7bf79f95496fd3041b527/src/mpi/romio/adio/ad_gpfs/ad_gpfs_rdcoll.c#L185 resulting in ranks calling either an Allreduce or a bunch of Allgathers.

I think it only happens when the environment is not PE or BG (but can't test that), so possibly also relates to #2307?
My impression is it's due to the lines https://github.com/pmodels/mpich/blob/2135f96a0580148563f7bf79f95496fd3041b527/src/mpi/romio/adio/ad_gpfs/ad_gpfs.c#L28-L32 Some of the tuning parameters for GPFS are set differently on (non-)aggregator ranks because not all call `ad_gpfs_get_env_vars()`, which would presumably only happen inside `ADIOI_GPFS_SetInfo`?
Default values for those tuning parameters are `!=0`, but ranks not calling the initialisation function will have value `0`, so end up in another branch of some of the `if` conditions?
Manually setting the tuning parameters like to zero helps in not triggering the deadlock, but limits tunability.

My simple-minded attempt of adding a call to `ad_gpfs_get_env_vars()` to the SetInfo function did not seem to work. Maybe it needs to go to another place? I don't really know how all of this is meant to play together.

**Comment by dbroemmel** [comment]
Ok, it doesn't help changing the SetInfo function because this one isn't called. Since we never had problems on BlueGene I assumed the different behaviour is due to a platform switch and the easiest I found was the one quoted above.
I did some more debugging and now think the divergence happens e.g. here
https://github.com/pmodels/mpich/blob/61f175301ed84aeff014c2faefe17930c52c4b1d/src/mpi/romio/adio/common/ad_opencoll.c#L73-L93 where not all ranks will call `ADIOI_GPFS_Open` which in turn calls `ad_gpfs_get_env_vars()`.

Forcing `ad_gpfs_get_env_vars()` to be called in the collective read operation removes the deadlock.

Since I'm still guessing, I'll leave the proper fix to the experts.
